### DISCUSSION TITLE
Records should allow attributes on the primary constructor

### DISCUSSION THREAD

**Discussion by RikkiGibson** [dotnet/csharplang#3650] [Language Ideas]
Slightly related to dotnet/roslyn#44691

Currently there is no way to declare attributes on the primary constructor of a record.

```cs
[Attr] // ... no target or distinct syntax exists for the primary constructor
record Rec(int X, int Y);
```

If we decide to allow declaring an explicit "body" for a primary constructor, we could make it so attributes may be applied to that syntax which solely corresponds to the constructor.

```cs
record Rec(int X, int Y)
{
    data int Z;

    // strawman "body" syntax with an attribute:
    [Attr]
    {
        Z = X + Y;
    }
}
```

**Comment by HaloFour** [comment]
Sounds like we need more attribute targets?

```cs
[constructor: Attr]
public record Rec(
    [property: Foo] int X,
    [field: NonSerialized] int Y
);
```

**Reply by raffaeler** [reply]
I second that because the intent is more clear in comparison to other proposals.

The practical use-case I had is when deserializing with `System.Text.Json` : https://github.com/dotnet/runtime/issues/45373 
I believe this issue will become more and more popular.

**Comment by 333fred** [comment]
We had a discussion on the internal mailing list, and the sentiment seemed to be preferring `method` as the target. It's already the target you use on constructors today:

```cs
 using System;
[AttributeUsage(AttributeTargets.Constructor)]
public class Attr : Attribute{}
public class C {
    [method: Attr]
    public C()
    {
    }
}
```

**Comment by bbarry** [comment]
Would it make sense that `[method:...]` attributes added to the class would get attached to all constructors in general?

edit to be clear:

```csharp
[method:Attr]
public class Foo { }
```

not just records. Perhaps to implicit constructors only?

**Comment by RikkiGibson** [comment]
I would not attach such attributes to any explicitly declared constructors within the record. One could argue that such attributes should be attached to the synthesized copy constructor as well as the primary constructor, though. I personally do not see utility in this because I see the copy constructor as an implementation detail of with-expressions that users should not need to customize.

**Reply by Rekkonnect** [reply]
I believe there should be some way to attach attributes to the synthesized copy ctor, even if it means adding a new attribute target

**Reply by 333fred** [reply]
> I believe there should be some way to attach attributes to the synthesized copy ctor

For what purpose? That constructor is not called by user code.

**Reply by raffaeler** [reply]
> For what purpose? That constructor is not called by user code.

An example is the `System.Text.Json` library, which uses attributes to identify the ctor that should be used to deserialize an object.
Currently, there is no way to instruct the deserializer to use the default ctor to deserialize a json document and this means avoiding records completely in such scenario. See https://github.com/dotnet/runtime/issues/45373

**Reply by CyrusNajmabadi** [reply]
why would S.T.J need to call the synthesized copy constructor?

**Reply by raffaeler** [reply]
> why would S.T.J need to call the synthesized copy constructor?

As I wrote in the provided link: «When a record declares other constructors in addition to the primary one, an exception is thrown and there is no way to put the JsonConstructor attribute on the primary.»

I believe it would be better for the dev to choose which ctor the library should use instead of throwing or assuming one to be the best choice.

**Reply by CyrusNajmabadi** [reply]
@raffaeler I don't see what that has to do with teh synthesized copy constructor though.  The statement was: `I believe there should be some way to attach attributes to the synthesized copy ctor`.  We're looking for a justification for why that constructor needs attributes.

**Reply by raffaeler** [reply]
Sorry @CyrusNajmabadi, I missed it was about the **copy** ctor. My fault.

**Reply by 333fred** [reply]
Right. That constructor is a protected constructor that is (usually) only called by the synthesized clone virtual method used by `with`. I'm not sure why that ctor would need to be annotated with any info.

**Comment by qwertie** [comment]
I think I need this feature to fix an exception I'm getting from System.Text.Json on a record type with secondary constructors: 

System.NotSupportedException: Deserialization of types without a parameterless constructor, a singular parameterized constructor, or a parameterized constructor annotated with 'JsonConstructorAttribute' is not supported.

**Reply by ugurpelister** [reply]
Exact same problem with MessagePack deserialization.

**Comment by DrPhil** [comment]
Is there a github issue or anything where we can track this, or is this discussion the official place for getting news about implementation status?

**Reply by HaloFour** [reply]
If this proposal becomes "championed" an issue will be opened to represent it and it would be linked to from this discussion.  That doesn't seem to have happened yet.

**Comment by drieseng** [comment]
Unless I'm missing something, it's the same issue all over again with primary constructors. I started a new discussion for this [here](https://github.com/dotnet/csharplang/discussions/7019).

**Reply by RikkiGibson** [reply]
Thanks for pinging on this. I'd like to treat this discussion as an umbrella for both records and non-records. It will help us to be able to gather the feedback, reactions etc in one place.

**Comment by RikkiGibson** [comment]
I will champion this. I just don't have a proposal yet.

**Reply by poizan42** [reply]
See #7047
